src/Microsoft.AspNetCore.Sockets.Client/WebSocketsTransport.cs

```diff
@@ -0,0 +1,139 @@
+            {
+                throw new ArgumentNullException(nameof(application));
+            }
+            _application = application;
+            await Connect(url);
+            var sendTask = SendMessages(url, _cancellationToken);
+            var receiveTask = ReceiveMessages(url, _cancellationToken);
+
+            Running = Task.WhenAll(sendTask, receiveTask).ContinueWith(t => {
+                _application.Output.TryComplete(t.IsFaulted ? t.Exception.InnerException : null);
+                return t;
+            }).Unwrap();
+        }
+
+        private async Task ReceiveMessages(Uri pollUrl, CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                const int bufferSize = 1024;
+                var totalBytes = 0;
+                var incomingMessage = new List<byte[]>();
+                WebSocketReceiveResult receiveResult;
+                do
+                {
+                    var buffer = new ArraySegment<byte>(new byte[bufferSize]);
+                    receiveResult = await _webSocket.ReceiveAsync(buffer, cancellationToken);
+                    incomingMessage.Add(buffer.Array);
+                    totalBytes += receiveResult.Count;
+                } while (!receiveResult.EndOfMessage);
+
+                Message message;
+                if (incomingMessage.Count > 1)
+                {
+                    var messageBuffer = new byte[totalBytes];
+                    var offset = 0;
+                    for (var i = 0 ; i < incomingMessage.Count; i++)
+                    {
+                        Buffer.BlockCopy(incomingMessage[i], 0, messageBuffer, offset, incomingMessage[i].Length);
+                        offset += incomingMessage[i].Length;
+                    }
+                    
+                    message = new Message(ReadableBuffer.Create(messageBuffer).Preserve(), receiveResult.MessageType == 0 ? Format.Text : Format.Binary, receiveResult.EndOfMessage);
+                }
+                else
+                {
+                    message = new Message(ReadableBuffer.Create(incomingMessage[0]).Preserve(), receiveResult.MessageType == 0 ? Format.Text : Format.Binary, receiveResult.EndOfMessage);
+                }
+
+                while (await _application.Output.WaitToWriteAsync(cancellationToken))
+                {
+                    if (_application.Output.TryWrite(message))
+                    {
+                        incomingMessage.Clear();
+                        break;
+                    }
+                }
+            }
+        }
+
+        private async Task SendMessages(Uri sendUrl, CancellationToken cancellationToken)
+        {
+            while (await _application.Input.WaitToReadAsync(cancellationToken))
+            {
+                Message message;
+                while (!cancellationToken.IsCancellationRequested && _application.Input.TryRead(out message))
```

Comment:
I don't like this extra condition, let the thing throw an operationcancelledexception